build: update go to 1.16.5

rail opened the pull request:
Fixes #66404

* Upgrade gazelle to latest version. Go 1.16 updates `go.sum` when
  gazelle calls `go download`. The new version fixes this issue. See
  https://github.com/bazelbuild/bazel-gazelle/pull/1015.
* Remove line references in `build/README.md`, because they are not
  stable.
* Set `GOPATH` in various places for bazel based builds and tests in
  order to work round the case when `go env` fails when `GOPATH` and
  `HOME` are unset. See https://github.com/golang/go/issues/43938 for
  the details.
* Set `GO111MODULE=off` in `pkg/acceptance/compose/gss/psql/Dockerfile`.
  Go 1.16 requires `go.mod` in order to build and test the module
  without `GO111MODULE=off`.

Checklist:

* [x] Adjust the Pebble tests to run in new version.
* [x] Adjust version in Docker image ([source](./builder/Dockerfile)).
* [x] Adjust version in the TeamCity agent image ([setup script](./packer/teamcity-agent.sh))
* [x] Rebuild and push the Docker image (following [Basic Process](#basic-process))
* [x] Bump the version in `WORKSPACE` under `go_register_toolchains`. You may need to bump [rules_go](https://github.com/bazelbuild/rules_go/releases).
* [x] Bump the version in `builder.sh` accordingly ([source](./builder.sh#L6)).
* [x] Bump the version in `go-version-check.sh` ([source](./go-version-check.sh)), unless bumping to a new patch release.
* [x] Bump the go version in `go.mod`. You may also need to rerun `make vendor_rebuild` if vendoring has changed.
* [x] Bump the default installed version of Go in `bootstrap-debian.sh` ([source](./bootstrap/bootstrap-debian.sh)).
* [x] Replace other mentions of the older version of go (grep for `golang:<old_version>` and `go<old_version>`).
* [x] Update the `builder.dockerImage` parameter in the TeamCity [`Cockroach`](https://teamcity.cockroachdb.com/admin/editProject.html?projectId=Cockroach&tab=projectParams) and [`Internal`](https://teamcity.cockroachdb.com/admin/editProject.html?projectId=Internal&tab=projectParams) projects.
* [ ] Ask the Developer Infrastructure team to deploy new TeamCity agent images according to [packer/README.md](./packer/README.md)

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/66505)
<!-- Reviewable:end -->

rail committed:
build: update go to 1.16.5

Fixes #66404

* Upgrade gazelle to latest version. Go 1.16 updates `go.sum` when
  gazelle calls `go download`. The new version fixes this issue. See
  https://github.com/bazelbuild/bazel-gazelle/pull/1015.
* Remove line references in `build/README.md`, because they are not
  stable.
* Set `GOPATH` in various places for bazel based builds and tests in
  order to work round the case when `go env` fails when `GOPATH` and
  `HOME` are unset. See https://github.com/golang/go/issues/43938 for
  the details.
* Set `GO111MODULE=off` in `pkg/acceptance/compose/gss/psql/Dockerfile`.
  Go 1.16 requires `go.mod` in order to build and test the module
  without `GO111MODULE=off`.

Checklist:

* [x] Adjust the Pebble tests to run in new version.
* [x] Adjust version in Docker image ([source](./builder/Dockerfile)).
* [x] Adjust version in the TeamCity agent image ([setup script](./packer/teamcity-agent.sh))
* [x] Rebuild and push the Docker image (following [Basic Process](#basic-process))
* [x] Bump the version in `WORKSPACE` under `go_register_toolchains`. You may need to bump [rules_go](https://github.com/bazelbuild/rules_go/releases).
* [x] Bump the version in `builder.sh` accordingly ([source](./builder.sh#L6)).
* [x] Bump the version in `go-version-check.sh` ([source](./go-version-check.sh)), unless bumping to a new patch release.
* [x] Bump the go version in `go.mod`. You may also need to rerun `make vendor_rebuild` if vendoring has changed.
* [x] Bump the default installed version of Go in `bootstrap-debian.sh` ([source](./bootstrap/bootstrap-debian.sh)).
* [x] Replace other mentions of the older version of go (grep for `golang:<old_version>` and `go<old_version>`).
* [ ] Update the `builder.dockerImage` parameter in the TeamCity [`Cockroach`](https://teamcity.cockroachdb.com/admin/editProject.html?projectId=Cockroach&tab=projectParams) and [`Internal`](https://teamcity.cockroachdb.com/admin/editProject.html?projectId=Internal&tab=projectParams) projects.
* [ ] Ask the Developer Infrastructure team to deploy new TeamCity agent images according to [packer/README.md](./packer/README.md)

Release note: None

otan commented:
![image](https://user-images.githubusercontent.com/3646147/122507997-9751b500-d044-11eb-9863-8023208fc61c.png)

congrats! always a doozy.

rail commented:
> bazel made it worse haha

I'm totally fine with this, given how it improves reproducibility. \o/

rickystewart commented on build/STRINGER.bzl:
```diff
@@ -12,7 +12,8 @@ def stringer(src, typ, name):
          GO_REL_PATH=`dirname $(location @go_sdk//:bin/go)`
          GO_ABS_PATH=`cd $$GO_REL_PATH && pwd`
-         env PATH=$$GO_ABS_PATH HOME=$(GENDIR) \
+         # Set GOPATH to something to workaround https://github.com/golang/go/issues/43938
```
Does this break the Windows build?

rail replied:
This is a good  question... I haven't explicitly checked it. I hope it just works, because we use Unix slashes in the same script.

rickystewart replied:
The build passed, you're good :)

rickystewart commented on pkg/cmd/urlcheck/BUILD.bazel:
```diff
@@ -5,7 +5,7 @@ go_library(
     importpath = "github.com/cockroachdb/cockroach/pkg/cmd/urlcheck",
     visibility = ["//visibility:private"],
-    deps = ["//pkg/cmd/urlcheck/lib/urlcheck"],
+    deps = ["@com_github_cockroachdb_cockroach//pkg/cmd/urlcheck/lib/urlcheck:go_default_library"],
```
What's going on here? If Gazelle inserted this, it seems wrong.

```
bazel query @com_github_cockroachdb_cockroach//...
ERROR: No such repository '@com_github_cockroachdb_cockroach'
ERROR: no targets found beneath ''
Loading: 0 packages loaded
```

rickystewart replied:
i.e., the original `dep` was good, this definitely shouldn't have been changed

rail replied:
That was gazelle. Let me poke it more to see the exact issue. Without this change the lint test was failing.

rickystewart replied:
You can probably do a `# gazelle:resolve go github.com/cockroachdb/cockroach/pkg/cmd/urlcheck/lib/urlcheck //pkg/cmd/urlcheck/lib/urlcheck` in `BUILD.bazel` (or similar, don't remember the exact syntax) to quash this.

rail replied:
Thank you for the tip. It just worked fine. \o/

rail commented:
bors r=rickystewart

rail commented:
Flaky tests!!!

rail commented:
bors r=rickystewart
